packages/jest-mock/src/index.ts

```diff
@@ -38,33 +26,65 @@ type MockFunctionResult = {
   /**
    * Indicates how the call completed.
    */
-  type: MockFunctionResultType,
+  type: MockFunctionResultType;
   /**
    * The value that was either thrown or returned by the function.
    * Undefined when type === 'incomplete'.
    */
-  value: any,
+  value: unknown;
 };
 
-type MockFunctionState = {
-  instances: Array<any>,
-  calls: Array<Array<any>>,
+type MockFunctionState<T, Y extends unknown[]> = {
+  calls: Array<Y>;
+  instances: Array<T>;
+  invocationCallOrder: Array<number>;
   /**
    * List of results of calls to the mock function.
    */
-  results: Array<MockFunctionResult>,
-  invocationCallOrder: Array<number>,
+  results: Array<MockFunctionResult>;
 };
 
 type MockFunctionConfig = {
-  isReturnValueLastSet: boolean,
-  defaultReturnValue: any,
-  mockImpl: any,
-  mockName: string,
-  specificReturnValues: Array<any>,
-  specificMockImpls: Array<any>,
+  isReturnValueLastSet: boolean;
+  defaultReturnValue: unknown;
+  mockImpl: Function | undefined;
+  mockName: string;
+  specificReturnValues: Array<unknown>;
+  specificMockImpls: Array<Function>;
 };
 
+interface Mock<T, Y extends unknown[] = unknown[]>
+  extends Function,
+    MockInstance<T, Y> {
+  new (...args: Y): T;
+  (...args: Y): T;
+}
+
+interface SpyInstance<T, Y extends unknown[]> extends MockInstance<T, Y> {}
+
+interface MockInstance<T, Y extends unknown[]> {
+  _isMockFunction: boolean;
+  _protoImpl: Function;
+  getMockName(): string;
+  getMockImplementation(): Function | undefined;
+  mock: MockFunctionState<T, Y>;
+  mockClear(): void;
+  mockReset(): void;
+  mockRestore(): void;
+  mockImplementation(fn: (...args: Y) => T): Mock<T, Y>;
```

Comment:
I think we can make these return types `this` to be more accurate (and shorter)?